### DISCUSSION TITLE
TissueMicroarrayExtractor

### DISCUSSION THREAD

**Discussion by mhoibo** [FAST-Imaging/FAST#204] [Q&A] [answered]
Hi, I have had some issues with the TissueMicroarrayExtractor. I have previously used it on TMA slides where it worked well, and running the extractor on those TMAs still works. 

However, now I have TMA scans from a new scanner, and it does not work as well anymore. From a single TMA slide including more than a 100 TMA cores the extractor only extracts 10-15 cores, and they are all only extractions of part of the core. None of the most solid/full cores are extracted,

I tested running the fast.TissueSegmentation on some of the new scans, and the segmentations are not as good (though many of them still are alright). It might seems like the default threshold does not segment as well on those images, but there are also examples of stroma that is not segmented at all, even with a quite strong color, and some strange circular shapes in the segmentations. In general the segmenations are a lot less like full circles, and more broken up. There are also quite a few tiny tissue artefacts around in some of the slides (though I though they are too small to be counted when calculating mean and median area?). Many of the semi-large parts of a TMA core when the segmentation is incomplete might be large enough though, yielding an incorrect mean and median region area. 

Looking at the segmentations alone did this:

        importer = fast.WholeSlideImageImporter.create(path/to/image)
        
        seg = fast.TissueSegmentation.create().connect(importer)

        renderer = fast.ImagePyramidRenderer.create().connect(importer)
        segmentationRenderer = fast.SegmentationRenderer.create().connect(seg)

        fast.SimpleWindow2D.create() \
            .connect(renderer) \
            .connect(segmentationRenderer) \
            .run()

Extracting TMA cores I did this:

       extractor = fast.TissueMicroArrayExtractor.create(level=args.level).connect(importer)
       for i, core in enumerate(fast.DataStream(extractor)):      
        ...

Is there a good way of using the tma-extractor for different slides, if you think the problem is related to the segmentation? In a way where on does not have to change parameters manually between slides/scanners?

 I can also see slight differences in color on the slides from the two scanners. 

Thank you in advance!

**Comment by smistad** [comment]
Hi @mhoibo 

If TissueSegmentation fails, then the TissueMicroArrayExtractor will not work either. 
Have you tried lowering the threshold parameter in TissueSegmentation? Does that give a more correct segmentation?
If so, we can add this parameter to the TissueMicroArrayExtractor as well.

**Reply by mhoibo** [reply]
Hi, yes, I tried the tissue segmentation alone, with a low threshold-value to test, and that gave a better segmentation. Then only very fatty areas are not segmented very well.

**Reply by smistad** [reply]
Ok, is the segmentation good enough then? 
I will add the threshold parameter to the TissueMicroArrayExtractor soon.

**Reply by andreped** [reply]
> Hi, yes, I tried the tissue segmentation alone, with a low threshold-value to test, and that gave a better segmentation. Then only very fatty areas are not segmented very well.

This is a well-known problem in histopathological image analysis. Sure you can tweak parameters to get better tissue segmentation for images from one scanner, but it will never generalise to all future cases. That's why people have started training neural networks to segment all tissue instead, or just have one method that segments all fatty tissue.

Then again, for this use case, it might be helpful to also reduce the morphological post-processing, which at least should then segment fatty tissue better without removing fragments you do not want to remove. See here:
https://fast.eriksmistad.no/classfast_1_1TissueSegmentation.html

Perhaps try setting dilationSize and erosionSize to 0 and see what happens. With a tuned threshold it should likely segment fatty tissue better.

**Reply by mhoibo** [reply]
Yes, I think for the use-case I have now, I can find a threshold that is good enough. Thank you, @smistad! I tried to to tune dilation and erosion too, and it might help to give better results, but will have to play around a bit to find good values:)

**Reply by smistad** [reply, marked as answer]
I have added the parameters to the TissueMicroArrayExtractor in the latest commit: https://github.com/smistad/FAST/actions/runs/7799787880

We are working on a neural network for improved tissue segmentation, the plan is to make that available in FAST to replace the current color thresholding method.

**Reply by mhoibo** [reply]
Thank you!
